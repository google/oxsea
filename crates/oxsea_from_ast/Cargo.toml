[package]
name = "oxsea_from_ast"
version = "0.1.0"
edition = "2021"

[dependencies]
oxc_allocator.workspace = true
oxc_ast.workspace = true
oxc_index.workspace = true
oxc_semantic.workspace = true

oxsea_ir = { path = "../oxsea_ir" }

[dev-dependencies]
oxc_codegen.workspace = true
oxc_parser.workspace = true
