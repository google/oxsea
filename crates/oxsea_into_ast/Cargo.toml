[package]
name = "oxsea_into_ast"
version = "0.1.0"
edition = "2021"

[dependencies]
oxc_allocator.workspace = true
oxc_ast.workspace = true

oxsea_ir = { path = "../oxsea_ir" }

[dev-dependencies]
oxc_codegen.workspace = true
