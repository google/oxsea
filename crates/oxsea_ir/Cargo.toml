[package]
name = "oxsea_ir"
version = "0.1.0"
edition = "2021"

[dependencies]
oxc_ast.workspace = true
