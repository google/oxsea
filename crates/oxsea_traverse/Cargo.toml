[package]
name = "oxsea_traverse"
version = "0.1.0"
edition = "2021"

[dependencies]
oxc_ast.workspace = true

oxsea_ir = { path = "../oxsea_ir" }
