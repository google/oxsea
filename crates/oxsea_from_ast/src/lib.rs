// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::{collections::HashMap, panic};

use oxc_ast::{
    ast::{AssignmentTarget, BinaryOperator, BindingPatternKind, Program, TSModuleDeclaration},
    visit::walk::walk_program,
    AstKind, Visit,
};
use oxc_index::IndexVec;
use oxc_semantic::{ReferenceFlags, Semantic, SemanticBuilder, SymbolId};
use oxsea_ir::{CompileTimeValue, IRGraph, IRNodeId, IR_END_ID, IR_START_ID};

struct FromAST<'i> {
    ir: IRGraph,
    _program: &'i Program<'i>,
    _semantic: &'i Semantic<'i>,

    control_tail: IRNodeId,
    value_stack: Vec<IRNodeId>,

    symbol_table: IndexVec<SymbolId, IRNodeId>,
    named_exports: HashMap<SymbolId, String>,
}

impl<'a> Visit<'a> for FromAST<'a> {
    fn leave_node(&mut self, kind: AstKind<'a>) {
        match kind {
            AstKind::NumericLiteral(l) => {
                self.value_stack
                    .push(self.ir.add_constant(CompileTimeValue::Number(l.value)));
            }
            AstKind::BooleanLiteral(l) => {
                self.value_stack
                    .push(self.ir.add_constant(CompileTimeValue::Boolean(l.value)));
            }
            AstKind::BinaryExpression(b) => match b.operator {
                oxc_ast::ast::BinaryOperator::Addition => {
                    let right = self.value_stack.pop().unwrap();
                    let left = self.value_stack.pop().unwrap();
                    self.value_stack.push(self.ir.add_add(left, right));
                }
                BinaryOperator::StrictEquality | BinaryOperator::Equality => {
                    let right = self.value_stack.pop().unwrap();
                    let left = self.value_stack.pop().unwrap();
                    self.value_stack
                        .push(self.ir.add_compare(left, right, b.operator));
                }
                _ => {
                    panic!(
                        "BinaryExpression with unsupported operator: {:?}",
                        b.operator
                    );
                }
            },
            AstKind::ReturnStatement(_) => {
                let value = self.value_stack.pop().unwrap();
                self.control_tail = self.ir.add_return(self.control_tail, value);
            }
            AstKind::VariableDeclaration(v) => {
                // Ignore declare-only cases, they only matter for type checking.
                if v.declare {
                    return;
                }

                // This is a severe oversimplification of what's _actually_ required.
                if v.declarations.len() != 1 {
                    panic!("VariableDeclaration with multiple declarations");
                }
                for decl in &v.declarations {
                    if decl.init.is_none() {
                        panic!("VariableDeclaration without init, not handled yet");
                    }

                    match &decl.id.kind {
                        BindingPatternKind::BindingIdentifier(b) => {
                            let symbol = b.symbol_id();
                            let value = self.value_stack.pop().unwrap();
                            self.symbol_table[symbol] = value;
                        }
                        _ => {
                            panic!("VariableDeclaration with non-binding identifier");
                        }
                    }
                }
            }
            AstKind::IdentifierReference(r) => {
                let reference = self._semantic.symbols().get_reference(r.reference_id());
                if reference.flags().contains(ReferenceFlags::Read) {
                    if let Some(symbol_id) = reference.symbol_id() {
                        let mut value = self.symbol_table[symbol_id];
                        if value == 0 {
                            let symbol_name = self._semantic.symbols().get_name(symbol_id);
                            value = self.ir.add_load_global(symbol_name.to_string());
                        }
                        self.value_stack.push(value);
                    } else {
                        // This would happen when reading an undeclared global variable.
                        let value = self.ir.add_load_global(r.name.to_string());
                        self.value_stack.push(value);
                    }
                }
            }
            AstKind::AssignmentExpression(a) => {
                if !a.operator.is_assign() {
                    panic!("AssignmentExpression with non-assign operator");
                }
                let value = self.value_stack.pop().unwrap();
                match &a.left {
                    AssignmentTarget::AssignmentTargetIdentifier(r) => {
                        let reference = self._semantic.symbols().get_reference(r.reference_id());
                        if let Some(symbol) = reference.symbol_id() {
                            self.symbol_table[symbol] = value;
                        } else {
                            panic!("AssignmentExpression with no symbol");
                        }
                    }
                    _ => {
                        panic!("AssignmentExpression with non-binding identifier");
                    }
                }
            }
            AstKind::ExportNamedDeclaration(e) => {
                // Ignore non-value exports, they are for type checking only.
                if !e.export_kind.is_value() {
                    return;
                }

                // Find all symbol ids that are being declared.
                if let Some(decl) = &e.declaration {
                    match decl {
                        oxc_ast::ast::Declaration::VariableDeclaration(v) => {
                            for decl in &v.declarations {
                                match &decl.id.kind {
                                    BindingPatternKind::BindingIdentifier(b) => {
                                        let symbol = b.symbol_id();
                                        self.named_exports.insert(symbol, b.name.to_string());
                                    }
                                    _ => {
                                        panic!("VariableDeclaration with non-binding identifier");
                                    }
                                }
                            }
                        }
                        _ => {
                            panic!("ExportNamedDeclaration with non-VariableDeclaration");
                        }
                    }
                }
            }
            AstKind::ExportDefaultDeclaration(_) => {
                let value = self.value_stack.pop().unwrap();
                self.ir
                    .add_bind_export(self.control_tail, "default".to_string(), value);
            }
            _ => { /* Ignore other nodes. */ }
        }
    }

    fn visit_ts_module_declaration(&mut self, _it: &TSModuleDeclaration<'a>) {
        /* Ignore contents of `declare ...` blocks. */
    }

    fn visit_if_statement(&mut self, it: &oxc_ast::ast::IfStatement<'a>) {
        let control_input = self.control_tail;
        // First, collect the node for our test expression.
        self.visit_expression(&it.test);
        let condition = self.value_stack.pop().unwrap();
        let branch_id = self.ir.add_if_else(control_input, condition);
        let consequent_id = self.ir.add_proj(branch_id, 0);
        let alternate_id = self.ir.add_proj(branch_id, 1);

        let original_symbol_table = self.symbol_table.clone();

        self.control_tail = consequent_id;
        self.visit_statement(&it.consequent);
        let consequent_tail = self.control_tail;
        let consequent_symbol_table = self.symbol_table.clone();

        let mut alternate_tail = alternate_id;
        self.symbol_table = original_symbol_table;
        if let Some(alternate) = &it.alternate {
            self.control_tail = alternate_id;
            self.visit_statement(alternate);
            alternate_tail = self.control_tail;
        }

        let merge_id = self
            .ir
            .add_merge(branch_id, &[consequent_tail, alternate_tail]);

        for (symbol, value) in self.symbol_table.iter_mut_enumerated() {
            let consequent_value = consequent_symbol_table[symbol];
            if consequent_value != *value {
                // If the symbol has diverged, create a phi node for it.
                let phi = self.ir.add_phi(merge_id, &[consequent_value, *value]);
                *value = phi;
            }
        }

        self.control_tail = merge_id;
    }
}

impl<'i> FromAST<'i> {
    pub fn new(program: &'i Program, semantic: &'i Semantic<'i>) -> Self {
        let mut symbol_table = IndexVec::new();
        symbol_table.resize(semantic.symbols().len(), 0);
        Self {
            ir: IRGraph::new(),
            _program: program,
            _semantic: semantic,

            control_tail: IR_START_ID,
            value_stack: vec![],

            symbol_table: symbol_table,
            named_exports: HashMap::new(),
        }
    }

    fn link_exports(&mut self) {
        let root_scope_id = self._semantic.scopes().root_scope_id();
        for (scope_id, symbol_id, _) in self._semantic.scopes().iter_bindings() {
            if scope_id != root_scope_id {
                continue;
            }
            if let Some(export_name) = self.named_exports.get(&symbol_id) {
                let value = self.symbol_table[symbol_id];
                self.ir
                    .add_bind_export(self.control_tail, export_name.clone(), value);
            }
        }
    }

    pub fn try_from_ast(&mut self) -> Result<(), &'static str> {
        walk_program(self, self._program);
        if self.value_stack.len() != 0 {
            return Err("Value stack is not empty");
        }
        self.link_exports();
        self.ir.add_edge(self.control_tail, IR_END_ID);
        Ok(())
    }
}

pub fn ir_from_ast<'a>(program: &Program) -> Result<IRGraph, &'static str> {
    let semantic_result = SemanticBuilder::new()
        .with_check_syntax_error(true)
        .with_cfg(true)
        .build(&program);
    if !semantic_result.errors.is_empty() {
        return Err("Analysis failed");
    }
    let semantic = semantic_result.semantic;

    let mut from_ast = FromAST::new(&program, &semantic);

    from_ast.try_from_ast()?;

    Ok(from_ast.ir)
}

#[cfg(test)]
mod tests {
    use oxc_allocator::Allocator;
    use oxc_ast::ast::SourceType;
    use oxc_parser::Parser;
    use oxsea_ir::{CompileTimeValue, IRNode, IR_END_ID};

    fn ir_from_source(source: &str) -> oxsea_ir::IRGraph {
        let allocator = Allocator::default();
        let ret = Parser::new(&allocator, source, SourceType::tsx()).parse();
        let ir = super::ir_from_ast(&ret.program).unwrap();
        check_graph(&ir);
        ir
    }

    #[test]
    fn default_export() {
        let actual = ir_from_source("export default 13;");
        assert_eq!(
            actual.get_node(2),
            &IRNode::new(
                oxsea_ir::IRInstruction::Constant(CompileTimeValue::Number(13.0)),
                vec![],
                vec![3],
            )
        );
        assert_eq!(
            actual.get_node(3),
            &IRNode::new(
                oxsea_ir::IRInstruction::BindExport("default".to_string()),
                vec![1, 2],
                vec![IR_END_ID],
            )
        );
    }

    fn check_graph(ir: &oxsea_ir::IRGraph) {
        // Check that all inputs and outputs are linked properly.
        for node_id in 0..ir.len() {
            let node = ir.get_node(node_id);
            for &input in node.inputs() {
                assert!(
                    ir.get_node(input).outputs().contains(&node_id),
                    "Node: #{} {:#?}\nIs missing from outputs of #{} {:#?}\n",
                    node_id,
                    node,
                    input,
                    ir.get_node(input)
                );
            }
            for &output in node.outputs() {
                assert!(
                    ir.get_node(output).inputs().contains(&node_id),
                    "Node: #{} {:#?}\nIs missing from inputs of #{} {:#?}\n",
                    node_id,
                    node,
                    output,
                    ir.get_node(output)
                );
            }
        }
    }

    #[test]
    fn pruned_branch() {
        ir_from_source("let x = 1; if (true) { x = 2; } export default x;");
    }

    #[test]
    fn it_needs_work() {
        let actual = ir_from_source("return 43 + 2;");
        assert_eq!(
            actual.get_node(2),
            &IRNode::new(
                oxsea_ir::IRInstruction::Constant(CompileTimeValue::Number(43.0)),
                vec![],
                vec![],
            )
        );
        assert_eq!(
            actual.get_node(3),
            &IRNode::new(
                oxsea_ir::IRInstruction::Constant(CompileTimeValue::Number(2.0)),
                vec![],
                vec![],
            )
        );
        assert_eq!(
            actual.get_node(4),
            &IRNode::new(
                oxsea_ir::IRInstruction::Constant(CompileTimeValue::Number(45.0)),
                vec![],
                vec![5],
            )
        );
        assert_eq!(
            actual.get_node(5),
            &IRNode::new(oxsea_ir::IRInstruction::Return, vec![1, 4], vec![0],)
        );
    }
}
