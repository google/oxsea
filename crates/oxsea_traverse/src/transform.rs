use oxsea_ir::{IRGraph, IRInstruction, IRNodeId, IR_END_ID, IR_INVALID_ID, IR_START_ID};

use crate::{
    sort::reverse_topological_sort, Add, BindExport, Compare, Constant, IfElse, LoadGlobal, Merge,
    Phi, Proj, Return,
};

pub trait Transform {
    fn visit_constant(&mut self, constant: &Constant, out: &mut IRGraph) -> IRNodeId {
        self.transform_constant(constant, out)
            .unwrap_or_else(|| out.add_constant(constant.value().clone()))
    }

    fn transform_constant(&mut self, _constant: &Constant, _out: &mut IRGraph) -> Option<IRNodeId> {
        None
    }

    fn visit_return(&mut self, ret: &Return, out: &mut IRGraph) -> IRNodeId {
        self.transform_return(ret, out)
            .unwrap_or_else(|| out.add_return(ret.control_id(), ret.value_id()))
    }

    fn transform_return(&mut self, _ret: &Return, _out: &mut IRGraph) -> Option<IRNodeId> {
        None
    }

    fn visit_bind_export(&mut self, bind_export: &BindExport, out: &mut IRGraph) -> IRNodeId {
        self.transform_bind_export(bind_export, out)
            .unwrap_or_else(|| {
                out.add_bind_export(
                    bind_export.control_id(),
                    bind_export.name().to_string(),
                    bind_export.value_id(),
                )
            })
    }

    fn transform_bind_export(
        &mut self,
        _bind_export: &BindExport,
        _out: &mut IRGraph,
    ) -> Option<IRNodeId> {
        None
    }

    fn visit_load_global(&mut self, load_global: &LoadGlobal, out: &mut IRGraph) -> IRNodeId {
        self.transform_load_global(load_global, out)
            .unwrap_or_else(|| out.add_load_global(load_global.name().to_string()))
    }

    fn transform_load_global(
        &mut self,
        _load_global: &LoadGlobal,
        _out: &mut IRGraph,
    ) -> Option<IRNodeId> {
        None
    }

    fn visit_compare(&mut self, compare: &Compare, out: &mut IRGraph) -> IRNodeId {
        self.transform_compare(compare, out).unwrap_or_else(|| {
            out.add_compare(compare.left_id(), compare.right_id(), *compare.operator())
        })
    }

    fn transform_compare(&mut self, _compare: &Compare, _out: &mut IRGraph) -> Option<IRNodeId> {
        None
    }

    fn visit_add(&mut self, add: &Add, out: &mut IRGraph) -> IRNodeId {
        self.transform_add(add, out)
            .unwrap_or_else(|| out.add_add(add.left_id(), add.right_id()))
    }

    fn transform_add(&mut self, _add: &Add, _out: &mut IRGraph) -> Option<IRNodeId> {
        None
    }

    fn visit_if_else(&mut self, if_else: &IfElse, out: &mut IRGraph) -> IRNodeId {
        self.transform_if_else(if_else, out)
            .unwrap_or_else(|| out.add_if_else(if_else.control_id(), if_else.condition_id()))
    }

    fn transform_if_else(&mut self, _if_else: &IfElse, _out: &mut IRGraph) -> Option<IRNodeId> {
        None
    }

    fn visit_proj(&mut self, proj: &Proj, out: &mut IRGraph) -> IRNodeId {
        self.transform_proj(proj, out)
            .unwrap_or_else(|| out.add_proj(proj.control_id(), proj.index()))
    }

    fn transform_proj(&mut self, _proj: &Proj, _out: &mut IRGraph) -> Option<IRNodeId> {
        None
    }

    fn visit_merge(&mut self, merge: &Merge, out: &mut IRGraph) -> IRNodeId {
        self.transform_merge(merge, out)
            .unwrap_or_else(|| match merge.resolved_index() {
                Some(value_index) => {
                    if merge.has_phi_outputs() {
                        out.add_resolved_merge(merge.branch_point_id(), *value_index)
                    } else {
                        merge.branch_point_id()
                    }
                }
                None => out.add_merge(merge.branch_point_id(), &merge.merge_input_ids()),
            })
    }

    fn transform_merge(&mut self, _merge: &Merge, _out: &mut IRGraph) -> Option<IRNodeId> {
        None
    }

    fn visit_phi(&mut self, phi: &Phi, out: &mut IRGraph) -> IRNodeId {
        self.transform_phi(phi, out)
            .unwrap_or_else(|| out.add_phi(phi.control_id(), &phi.value_ids()))
    }

    fn transform_phi(&mut self, _phi: &Phi, _out: &mut IRGraph) -> Option<IRNodeId> {
        None
    }
}

pub fn transform<T>(ir: &IRGraph, transform: &mut T) -> IRGraph
where
    T: Transform,
{
    let order = reverse_topological_sort(ir);
    let mut context = crate::Context::new_with_ids(ir);

    let mut out = IRGraph::new();

    for node_id in order {
        if node_id == IR_START_ID || node_id == IR_END_ID {
            continue;
        }

        let node = ir.get_node(node_id);
        if node.is_dead_control() {
            context.ids[node_id] = IR_INVALID_ID;
            continue;
        }

        let new_node_id = match node.instruction() {
            IRInstruction::Constant(value) => transform.visit_constant(
                &Constant {
                    ctx: &context,
                    node_id,
                    node,
                    value: value,
                },
                &mut out,
            ),
            IRInstruction::Return => transform.visit_return(
                &Return {
                    ctx: &context,
                    node_id,
                    node,
                },
                &mut out,
            ),
            IRInstruction::BindExport(name) => transform.visit_bind_export(
                &BindExport {
                    ctx: &context,
                    node_id,
                    node,
                    name,
                },
                &mut out,
            ),
            IRInstruction::LoadGlobal(name) => transform.visit_load_global(
                &LoadGlobal {
                    ctx: &context,
                    node_id,
                    node,
                    name,
                },
                &mut out,
            ),
            IRInstruction::Compare(op) => transform.visit_compare(
                &Compare {
                    ctx: &context,
                    node_id,
                    node,
                    operator: op,
                },
                &mut out,
            ),
            IRInstruction::Add => transform.visit_add(
                &Add {
                    ctx: &context,
                    node_id,
                    node,
                },
                &mut out,
            ),
            IRInstruction::IfElse(None) => transform.visit_if_else(
                &IfElse {
                    ctx: &context,
                    node_id,
                    node,

                    resolved: &None,
                },
                &mut out,
            ),
            IRInstruction::Proj(index) => transform.visit_proj(
                &Proj {
                    ctx: &context,
                    node_id,
                    node,

                    index: *index,
                },
                &mut out,
            ),
            IRInstruction::Merge(resolved_index) => transform.visit_merge(
                &Merge {
                    ctx: &context,
                    node_id,
                    node,

                    merge_input_index: 0,
                    resolved_index,
                },
                &mut out,
            ),
            IRInstruction::Phi => transform.visit_phi(
                &Phi {
                    ctx: &context,
                    node_id,
                    node,
                },
                &mut out,
            ),
            _ => {
                panic!("Unhandled instruction: {:?}", node.instruction());
            }
        };
        context.ids[node_id] = new_node_id;
    }

    out
}

#[cfg(test)]
mod tests {
    use super::*;
    use oxsea_ir::CompileTimeValue::*;
    use oxsea_ir::IR_START_ID;

    struct CopyTransform {}

    impl Transform for CopyTransform {}

    #[test]
    fn copy_add_with_constants() {
        let mut ir = IRGraph::new();
        let left = ir.add_constant(Number(1.0));
        let right = ir.add_constant(Number(2.0));
        let add = ir.add_add(left, right);
        ir.add_return(IR_START_ID, add);
        let mut copy_transform = CopyTransform {};
        let out = transform(&ir, &mut copy_transform);
        assert!(out.len() == 4);
    }
}
