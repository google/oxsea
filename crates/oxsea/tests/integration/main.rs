// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use oxc_allocator::Allocator;
use oxc_ast::ast::SourceType;
use oxc_codegen::{CodeGenerator, CodegenOptions};
use oxc_parser::Parser;
use oxsea_from_ast::ir_from_ast;
use oxsea_into_ast::ir_into_ast;
use oxsea_ir::{ir_to_dot, IRGraph};
use oxsea_traverse::transform::{transform, Transform};
use std::fs;

fn ir_from_source(source: &str) -> oxsea_ir::IRGraph {
    let allocator = Allocator::default();
    let ret = Parser::new(&allocator, source, SourceType::tsx()).parse();
    ir_from_ast(&ret.program).unwrap()
}

fn convert_and_print(graph: &IRGraph) -> String {
    let allocator = Allocator::default();
    let program = ir_into_ast(&allocator, graph, false).unwrap();
    let mut options = CodegenOptions::default();
    options.single_quote = true;
    let printed = CodeGenerator::new()
        .with_options(options.clone())
        .build(&program);
    printed.code
}

struct CopyTransform {}

impl Transform for CopyTransform {}

fn copy_ir(ir: &IRGraph) -> IRGraph {
    let mut copy_transform = CopyTransform {};
    transform(&ir, &mut copy_transform)
}

#[test]
fn main() {
    insta::glob!("fixtures/**/*.{js,ts,jsx,tsx}", |path| {
        let code = fs::read_to_string(path).unwrap();
        let name = path.file_stem().unwrap().to_str().unwrap();
        let ir = ir_from_source(&code);

        let dot_snapshot = format!("{}\n", ir_to_dot(&ir));
        insta::with_settings!({ snapshot_path => path.parent().unwrap(), prepend_module_to_snapshot => false, snapshot_suffix => "ir" }, {
            insta::assert_snapshot!(name, dot_snapshot);
        });

        let code_snapshot = convert_and_print(&ir);
        insta::with_settings!({ snapshot_path => path.parent().unwrap(), prepend_module_to_snapshot => false, snapshot_suffix => "out" }, {
            insta::assert_snapshot!(name, code_snapshot);
        });

        let ir_copy = copy_ir(&ir);
        let copy_dot_snapshot = format!("{}\n", ir_to_dot(&ir_copy));
        insta::with_settings!({ snapshot_path => path.parent().unwrap(), prepend_module_to_snapshot => false, snapshot_suffix => "copy" }, {
            insta::assert_snapshot!(name, copy_dot_snapshot);
        });
    });
}
