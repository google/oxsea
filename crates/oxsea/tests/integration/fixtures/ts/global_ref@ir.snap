---
source: crates/oxsea/tests/integration/main.rs
expression: dot_snapshot
input_file: crates/oxsea/tests/integration/fixtures/ts/global_ref.ts
snapshot_kind: text
---
digraph {
  fontname="Helvetica,Arial,sans-serif";
  node [fontname="Helvetica,Arial,sans-serif"];
  edge [fontname="Helvetica,Arial,sans-serif"];
  n0 [label="", shape=circle, style=filled, fillcolor="#fff", height=0.3, width=0.3, penwidth=5];
  n7 -> n0:i0;
  n1 -> n0:i1;
  n1 [shape=point, style=filled, fillcolor="#000", height=0.25, width=0.25];
  n2 [shape=box, label="global 'a'"] [style=rounded];
  n3 [shape=box, label="global 'b'"] [style=rounded];
  n4 [shape=record, label="{{<i0> left|<i1> right}|+}"] [style=rounded];
  n2 -> n4:i0;
  n3 -> n4:i1;
  n5 [shape=box, label="global 'c'"] [style=rounded];
  n6 [shape=record, label="{{<i0> left|<i1> right}|+}"] [style=rounded];
  n4 -> n6:i0;
  n5 -> n6:i1;
  n7 [shape=record, label="{{<i0> value}|export 'default'}"] [style=filled, fillcolor="#ffffaa"];
  n6 -> n7:i0;
}
